ui/app/services/api.ts

```diff
@@ -0,0 +1,185 @@
+/**
+ * Copyright (c) HashiCorp, Inc.
+ * SPDX-License-Identifier: BUSL-1.1
+ */
+
+import Service, { service } from '@ember/service';
+import {
+  Configuration,
+  RequestContext,
+  ResponseContext,
+  AuthApi,
+  IdentityApi,
+  SecretsApi,
+  SystemApi,
+} from '@hashicorp/vault-client-typescript';
+import config from '../config/environment';
+
+import type AuthService from 'vault/services/auth';
+import type NamespaceService from 'vault/services/namespace';
+import type ControlGroupService from 'vault/services/control-group';
+import type FlashMessageService from 'vault/services/flash-messages';
+import type { ApiError, ApiResponse, HeaderMap, XVaultHeaders } from 'vault/api';
+
+export default class ApiService extends Service {
+  @service('auth') declare readonly authService: AuthService;
+  @service('namespace') declare readonly namespaceService: NamespaceService;
+  @service declare readonly controlGroup: ControlGroupService;
+  @service declare readonly flashMessages: FlashMessageService;
+
+  #responseCache = new Map<string, ApiResponse>();
+
+  // -- Pre Request Middleware --
+  setLastFetch = async (context: RequestContext) => {
+    const { url } = context;
+    const isPolling = config.APP.POLLING_URLS.some((str) => url.includes(str));
+
+    if (!isPolling) {
+      this.authService.setLastFetch(Date.now());
+    }
+  };
+
+  getControlGroupToken = async (context: RequestContext) => {
+    const { url, init } = context;
+    const controlGroupToken = this.controlGroup.tokenForUrl(url);
+    let newUrl = url;
+    // if we have a Control Group token that matches the intendedUrl,
+    // unwrap it and return the unwrapped response as if it were the initial request
+    // to do this, we rewrite the request
+    if (controlGroupToken) {
+      const { token } = controlGroupToken;
+      const { headers } = this.buildHeaders({ token });
+      newUrl = '/v1/sys/wrapping/unwrap';
+      init.method = 'POST';
+      init.headers = headers;
+      init.body = JSON.stringify({ token });
+    }
+    return { url: newUrl, init };
+  };
+
+  setHeaders = async (context: RequestContext) => {
+    const { url, init } = context;
+    const headers = new Headers(init.headers);
+    // unauthenticated or clientToken requests should set the header in initOverrides
+    // unauthenticated value should be empty string, not undefined or null
+    if (!headers.has('X-Vault-Token')) {
+      headers.set('X-Vault-Token', this.authService.currentToken);
+    }
+    if (init.method === 'PATCH') {
+      headers.set('Content-Type', 'application/merge-patch+json');
+    }
+    // use initOverrides to set the namespace header to something other than path set in the namespace service
+    // for requests that must be made to root namespace pass empty string as value
+    const namespace = this.namespaceService.path;
+    if (!headers.has('X-Vault-Namespace') && namespace) {
+      headers.set('X-Vault-Namespace', namespace);
+    }
```

Comment:
I captured it in a comment but there is no concept of metadata with the new client so there isn't a direct mapping for Ember Data `adapterOptions`. This is a good thing from my perspective and fortunately the client provides an option to override request parameters like headers for example. I added a `buildHeaders` method to the service to help with this when overriding headers on individual requests.

Reply:
Ya it's nice to be explicit now that for a given request we need to override a header so it's easier to see what's going on and simplifies the global header logic in the service.